File: 16. Matrix Chain Multiplication.py
# 16. Matrix Chain Multiplication
# Topic: Matrix Chain Multiplication
# Description: Given dimensions array, compute minimal scalar multiplications using memoization.
# Input: n followed by n+1 integers (dimensions)
# Output: minimum cost

import sys
from functools import lru_cache

if __name__ == '__main__':
    data = sys.stdin.read().strip().split()
    if not data:
        sys.exit(0)
    it = iter(data)
    n = int(next(it))
    dims = [int(next(it)) for _ in range(n+1)]

    @lru_cache(maxsize=None)
    def mcm(i: int, j: int) -> int:
        if i == j:
            return 0
        best = 10**30
        for k in range(i, j):
            cost = mcm(i, k) + mcm(k+1, j) + dims[i-1]*dims[k]*dims[j]
            if cost < best:
                best = cost
        return best

    print(mcm(1, n))

# Memoization process started at mcm decorated with lru_cache.
# Time Complexity: O(n^3)
# Space Complexity: O(n^2)

File: 16. Matrix Chain Multiplication.cpp
// 16. Matrix Chain Multiplication
// Topic: Matrix Chain Multiplication
// Description: Given dimensions array, compute minimal scalar multiplications using memoization.
// Input: n (number of matrices), followed by n+1 dimensions
// Output: minimum cost

#include <bits/stdc++.h>
using namespace std;
using ll = long long;

ll mcm(int i, int j, const vector<int>& dims, vector<vector<ll>>& memo)
{
    if (i == j)
        return 0;
    if (memo[i][j] != -1)
        return memo[i][j];
    ll best = LLONG_MAX;
    for (int k = i; k < j; ++k)
    {
        ll left = mcm(i, k, dims, memo);
        ll right = mcm(k + 1, j, dims, memo);
        ll cost = left + right + (ll)dims[i - 1] * dims[k] * dims[j];
        best = min(best, cost);
    }
    memo[i][j] = best;
    return best;
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n;
    if (!(cin >> n))
        return 0;
    vector<int> dims(n + 1);
    for (int i = 0; i <= n; ++i) cin >> dims[i];
    vector<vector<ll>> memo(n + 1, vector<ll>(n + 1, -1));
    cout << mcm(1, n, dims, memo) << '\n';
}

// Memoization process started in mcm where memo[i][j] caches minimal cost for multiplying Ai..Aj.
// Time Complexity: O(n^3) due to k loop and O(n^2) states
// Space Complexity: O(n^2)
